uvicorn/_types.py

```diff
@@ -0,0 +1,87 @@
+import sys
+from typing import Optional, Union
+
+if sys.version_info < (3, 8):
+    from typing_extensions import Literal, TypedDict
+else:
+    from typing import Literal, TypedDict
+
+# SCOPES
+
+
+class ASGIDict(TypedDict):
+    version: str
+    spec_version: Optional[Union[Literal["2.0"], Literal["2.1"]]]
+
+
+class LifespanScope(TypedDict):
+    """
+    https://asgi.readthedocs.io/en/latest/specs/lifespan.html#scope
+    """
+
+    type: Literal["lifespan"]
+    asgi: ASGIDict
+
+
+# lifespan messages receive
+class LifespanReceiveStartup(TypedDict):
+    """
+    https://asgi.readthedocs.io/en/latest/specs/lifespan.html#startup-receive-event
+    """
+
+    type: Literal["lifespan.startup"]
+
+
+class LifeSpanReceiveShutdown(TypedDict):
+    """
+    https://asgi.readthedocs.io/en/latest/specs/lifespan.html#shutdown-receive-event
+    """
+
+    type: Literal["lifespan.shutdown"]
+
+
+LifespanReceiveMessage = Union[LifespanReceiveStartup, LifeSpanReceiveShutdown]
```

Comment:
> (Handy trick here - Literal accepts multiple literal values, which mypy interprets as a Union of those literals.)

super nice, I learnt something today.

> (I know we'd need an optional message attribute for startup.shutdown.failed, but since we don't handle it anyway this is the form we need.)

I left both the message and the `lifespan.shutdown.failed` since we need message anyway for the `lifespan.startup.failed`